### ISSUE TITLE
Image Renamer result preview before fetching all data + documentation update request

### ISSUE THREAD

**Issue by tipakA** [Bionus/imgbrd-grabber#2436]
**Is your feature request related to a problem? Please describe**

Hello, I'm new to the program.
Yesterday i downloaded around 3500 images merging some of my older dumps together, and i initially downloaded them to `%search%/%md5%.%ext%`. Today i realised that this way any sorting is completely lost, so i decided to rename all images to be in `%id%-%md5%.%ext%` format, thus preserving the order.

First issue is that there are 5200 images in the grabber's directory. To see how things work, i tried looking at the [renamer docs](https://bionus.github.io/imgbrd-grabber/docs/tools/image-renamer.html), but they say almost nothing. Since i wouldn't lose much, i decided to hit ok and wait for renamer to fetch all the data, which it eventually did.
However, the Destination filename was set to `%search%/%id%-%md5%.%ext%`, making whole operation useless, as it would now create 5200 directories due to `%search%/`.

Seeing this, i cancelled the rename, and tried to instead do it on subfolder basis instead of on entire lib folder. This posed an issue though, because renamer would always say that there are 0 images to rename.

In the meantime i also went to [filenames docs](https://bionus.github.io/imgbrd-grabber/docs/filename.html) to look for any tokens that might be of use here, finding nothing. (I then did go to github repo and after a bit of searching found [this commit](https://github.com/Bionus/imgbrd-grabber/commit/2125a7ea39b3bbd8ca1c59dcacc3b33ecd99d335) that adds 2 tokens that probably would be of use, but they are ~~still~~ not documented there.

I realized later on that the 0 image issue was due to me having `%search%/` in the *origin* filename, making renamer look into subdirectories from the path that i passed while there are none.

<hr>

**Describe the solution you'd like**

`Destination filename` should have some form of preview of how the output will look like given the input.
I am aware that my library is small and that i have the will to go folder-by-folder, but that will not be the case for everyone. I also am aware that there is no metadata to display *without* fetching, but fetching every single image info *just* to see how the output will look like is definitely *not* an ideal solution.

Maybe an "example output" or "preview" button to fetch few images to display how the result of renaming will look like?

Additionally, while i did realize why was i getting "0 images to rename" and was able to dig for `%directory%` token, i have 0 clue how to use it and how it look like. Updating [filenames docs](https://bionus.github.io/imgbrd-grabber/docs/filename.html) to include new tokens would be nice, and [renamer docs](https://bionus.github.io/imgbrd-grabber/docs/tools/image-renamer.html) should probably get some content there to at least describe how does it behave basing on input (given that it defaults origin name to filename taken from settings)

**Additional context**

Sorry for the ramble, I'm not the best at creating issues 😅

**Comment by tipakA** [comment]
Now that i took a second look, Origin filename doesn't really care about what the token is, as long as it looks like you need a subdir from the passed path.

From the description of the custom tokens, it does not sound like i can pass a custom token that has names of multiple directories, and even if, the token gets ignored anyway.

<br >
Seems like i will indeed have to just go dir by dir then

**Comment by yami-no-tusbas** [comment]
> Seems like i will indeed have to just go dir by dir then

Well not in all case, assuming you didn't make the mistake of rennaming using %search% since it'll return as a string the md5 (or id, can't remember) of the image. You can still use your existing folder structure. Assuming you used somethin like : 

> c:\somedir\searchtag\imagemd5.ext

You can use the renamer on the  "c:\somedir" folder; grabber will see the subfolder, and as long as your image origin filename is good, it'll grab the md5's

So basically you want theses settings : 

Folder : `C:\somedir`
Origin filename : `%md5%.%ext%`
Destination filename : `%directory%\%id%-%md5%.%ext%`

This will keep the existing structure and rename your images. And even if your image naming doesn't work, instead of getting md5 from filename, you can make grabber calculate the md5, should work fine as long as you didn't alter the images.

And even in the case you did fail and now have 5000+ folders, you can use conditionnal filename to "re-tag" them, but this is out of the scope for this answer.

**Comment by tipakA** [comment]
Thanks for commenting.

I did actually go in dir-by-dir way already (and no, i did not click continue on that 5200 dirs rename 😅).
Now that there are only 88 unrenamed files (those were removed from gelbooru since i made my own dumps in late 2019, so there is nothing to fetch) instead of 5200, i can play around with output.

Doing it the way you suggest:
<table>
<thead></thead>
<tbody>
	<tr><td>Folder: <code>N:\PICTURE\_H\GRABBER</code></td></tr>
	<tr><td>Origin filename: <code>%search%/%md5%.%ext%</code> <sub>(it doesn't really matter what's the first token here)</sub></td></tr>
	<tr><td>Destination filename: <code>%directory%/%id%-%md5%.%ext%</code></td></tr>
	</tbody>
</table>

It does seem to resolve correctly on 7.6.0: <sub>(2nd image shown here to show that it goes between subdirs, but has no resolve as it doesn't exist on source)</sub>
![image](https://user-images.githubusercontent.com/31581159/128369931-3c7d96db-31b2-49ac-a0eb-0224bcb746a5.png)

<hr>

> Well not in all case, assuming you didn't make the mistake of rennaming using %search% since it'll return as a string the md5 (or id, can't remember) of the image.

Well, while it is clear in the hindsight what the `%search%` is, as the renamer uses md5 from the filename to search for metadata (at least in the case of me having original names as `%md5%.%ext%` and using gelbooru), i initially assumed that it would use result of the file search - the result of searching for files to rename in the given folder. (This doesn't exactly explain why would it only contain path without filename, but hey; if there are *javascript tags*, there could be code that filters that out if there is a `/` after it 🤷‍♀️)

<hr>

Seems like it is possible to do, but then we go back to the fact that `%directory%` token which would be needed for this is not documented, and the fact that before renaming anything i would have to do 5200 fetches every single time i would want to see how the results would look like (maybe there's a need here for some data caching 👀? (or maybe there is some unexplained/undocumented option somewhere that would allow caching of those to use later 👀👀👀))

**Comment by yami-no-tusbas** [comment]
> Seems like it is possible to do, but then we go back to the fact that `%directory%` token which would be needed for this is not documented, and the fact that before renaming anything i would have to do 5200 fetches every single time i would want to see how the results would look like (maybe there's a need here for some data caching 👀? (or maybe there is some unexplained/undocumented option somewhere that would allow caching of those to use later 👀👀👀))

Well I was thinking about it, because renamming only using md5 and making query to the board is a bit inneficient, not even talking "if" t he board goes offline, or your connection fail during the process (wich has happen to me on multiples occasions).

So since grabber has a md5 database in sqlite format of downloaded image, I was t hinking, maybe It cound have a "ImagesInformations" database, that would keep track of the data of an image, for quick renaming without having to ask the board for each images data. Information like Artists, md5, id, copyright(s), and most common used could be included. And obviously in the case of "updating" the user could choose to use the online queries. But at least caes like yours when you just want to keep the structure and add the missing "id" would not require to go online. And the databse could be updated by for exemple downloading the same image again (by id+md5). So the database won't be obsolete too quickly. (Wont work for ppl like me because I don't download an image twice, but still, be a good thing).

In short, a "bonus" caching database wich save some basic token about the images like artist, md5, id, website, and so on, since when the image is downloaded we already have the chance to get thoses information by the board. It would improve rennaming plus add an option to do it "offline". Plus it is a good point for board wo doesn't like having a lot of request like e621 or the unmantionnable board.

**Comment by typhoon71** [comment]
I second the idea of an additional sqlite db with the image data, maybe optional - but this is more of a feature request.
It would be useful for quick renamning, even if the server image was deleted.
